### PR TITLE
Optional promises

### LINKED ISSUE #191: How to handle promises rejected from store.dispatch?
With the newest update store.dispatch returns a promise for offline actions. I thought about my options to avoid unhandled promise rejections and this is what I came up with so far:

```js
export const mapDispatchToProps = actions => dispatch => {
  const dispatchWithCatch = action => {
    const dispatched = dispatch(action)
    return typeof dispatched === 'object' &&
      typeof dispatched.then === 'function'
      ? dispatched.catch(() => {})
      : dispatched
  }
  return {
    actions: bindActionCreators(actions, dispatchWithCatch),
  }
}
```

I think this new behavior should be optional. Or is it already and I am missing something? And what are your approaches to this?

### PR DESCRIPTION
**Pull request by wacii**
Set `offlineConfig.returnPromises = true` to return promises from `dispatch()`.

Closes #191

~edit: Ah, this includes commits from another PR. I'll clean this up in a bit.~

### CODE REVIEW

**Comment by wacii** [comment]
Alright, now the PR should be ready.

**Comment by sorodrigo** [comment]
Great! I’ll make a new release tomorrow with the pending stuff

File: src/middleware.js [hunk 1]
```diff
@@ -47,5 +47,5 @@ export const createOfflineMiddleware = (config: Config) => (store: any) => (
     send(offlineAction, store.dispatch, config, offline.retryCount);
   }
 
-  return promise || result;
+  return config.returnPromises ? promise || result : result;
```
Comment (piranna) [inline src/middleware.js:50]: If you use `await` in your code, it doesn't matter if you return a Promise or a value. Other thing is, if you are doing this feature opt-in to preserve backwards compatibility...

**Comment by sorodrigo** [comment]
Could you update the promise example? I think it won't be as straighforward because you need to change the offline config. So to make your life easier, maybe just add another case to `store.js` that checks for `process.env.REACT_APP_OFFLINE_API === 'promise'`. And hide the promise example from the App component if this env doesn't match.

**Comment by wacii** [comment]
I had forgotten the example used the returned promise. Nice catch.

File: src/middleware.js (hunk 1 above)
Comment (wacii) [reply src/middleware.js:50]: I agree. This is for backwards compatibility for as long as we are treating the feature as experimental.

File: examples/web/client/src/components/MakeRequests.js [hunk 2]
```diff
@@ -3,18 +3,15 @@ import { connect } from 'react-redux';
...
     const { successCallback, errorCallback } = this.props;
-    let promise;
-    if (successCallback) {
-      promise = handler().then(successCallback);
-    }
-    if (errorCallback) {
-      promise.catch(errorCallback);
-    }
-    return promise || handler();
+    return handler()
```
Comment (sorodrigo) [inline examples/web/client/src/components/MakeRequests.js:12]: Way better! 👍

File: examples/web/client/src/components/MakeRequests.js [hunk 3]
```diff
@@ -3,18 +3,15 @@ import { connect } from 'react-redux';
...
+      .then(successCallback || noop)
+      .catch(errorCallback || noop);
```
Comment (piranna) [inline examples/web/client/src/components/MakeRequests.js:14]: `noop()` is not needed, if the arguments for `then()` or `catch()` are `null` or `undefined`, then they are not executed. It's part of the spec.

File: src/middleware.js (hunk 1 above)
Comment (piranna) [reply src/middleware.js:50]: Maybe open a new issue to take this in account?

File: examples/web/client/src/components/MakeRequests.js (hunk 3 above)
Comment (sorodrigo) [reply examples/web/client/src/components/MakeRequests.js]: so instead it would be `.then().catch()`?

Comment (wacii) [reply examples/web/client/src/components/MakeRequests.js:14]: Oh! I did not know that. Nifty.

File: src/middleware.js (hunk 1 above)
Comment (wacii) [reply src/middleware.js:50]: Good idea.

File: examples/web/client/src/components/MakeRequests.js (hunk 3 above)
Comment (piranna) [reply examples/web/client/src/components/MakeRequests.js:14]: Good you like it :-)

**Pull request merged**